packages/costa/src/runtime.ts

```diff
@@ -229,13 +245,37 @@ export class CostaRuntime {
     }
     return pkg;
   }
+  /**
+   * fetch and check if the package name is valid.
+   * @param name the plugin package readstream for npm package tarball.
+   * @param cwd the current working directory.
+   */
+  async fetchByStream(stream :Readable, cwd?: string): Promise<PluginPackage> {
+    const source: PluginSource = {
+      from: 'tarball',
+      name: 'uploadedPackage'
+    };
+    const pkg = await fetchPackageJsonFromTarballStream(stream);
+
+    try {
+      this.validPackage(pkg);
+      this.assignPackage(pkg, source);
+    } catch (err) {
+      if (process.env.NODE_ENV === 'test') {
```

Comment:
This might be disallowed to inject some for test environment :(

Reply:
Sorry, this seems to be introduced by me, how about creating a new function, and move the above there, so that we can keep only this one hacky part.

Reply:
OK.